File: ml-benchmark/synduce/constraints/constantlist/index_of.f
Inductive List = elt Int | cons {Int, List};

head = \xs: List. match xs with elt w -> w | cons {h, t} -> h end;

is_const = fix (
  \f: List -> Bool. \xs: List.
  match xs with
    elt x -> true
  | cons {h, t} -> and (== h (head t)) (f t)
  end
);

@Input w: Int;

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt x -> if == w x then 1 else 0
  | cons {h, t} -> if == w h then 1 else if == 0 (f t) then 0 else + 1 (f t)
  end
);


target = fix (
  \f: List -> Compress List. \xs: List.
  match xs with
    elt a -> elt a
  | cons {h, t} -> cons {h, t} /*Avoid the recursion on t*/
  end
);

main = \xs: List. if is_const xs then spec (target xs) else 0;

File: ml-benchmark/synduce/constraints/constantlist/index_of.ml
type list = Elt of int | Cons of int * list

let head xs =
  match xs with
  | Elt w -> w
  | Cons (h, t) -> h

let rec is_const xs =
  match xs with
  | Elt x -> true
  | Cons (h, t) -> (h = head t) && is_const t

let w = read_int ()

let rec spec xs =
  match xs with
  | Elt x -> if w = x then 1 else 0
  | Cons (h, t) -> if w = h then 1 else if spec t = 0 then 0 else 1 + spec t

let rec target xs =
  match xs with
  | Elt a -> Elt a
  | Cons (h, t) -> Cons (h, t)

let program xs =
  if is_const xs then spec (target xs) else 0